Test2 (#6)
couchbaselabs/docs-tooling: example/modules/ROOT/pages/cbl-android-p2p.adoc

= Integrate a Custom Built Listener
:page-aliases: learn/java-android-p2psync-custom.adoc

> ifdef::show_edition[:page-edition: Enterprise Edition]

Author: Automated review comment from Vale and OpenAI:

  ```suggestion
  ifdef::show_edition[:page-edition: Couchbase Server Enterprise Edition]
  ```
  
  Using these Vale rules:
  * Couchbase.CouchbaseServerEnterprise (error) - Use the full product name (Couchbase Server Enterprise Edition) first.

> ifdef::show_edition[:page-edition: Enterprise Edition]

Author: Should probably be "Sync Gateway Enterprise Edition"? 🤔

> ifdef::show_edition[:page-edition: Enterprise Edition]

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  ifdef::show_edition[:page-edition: Couchbase Server Enterprise Edition]
  ```
  
  Using these Vale rules:
  * Couchbase.CouchbaseServerEnterprise (error) - Use the full product name (Couchbase Server Enterprise Edition) first.

> ifdef::show_edition[:page-edition: Enterprise Edition]

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  ifdef::show_edition[:page-edition: Couchbase Server Enterprise Edition]
  ```
  
  Using these Vale rules:
  * Couchbase.CouchbaseServerEnterprise (error) - Use the full product name (Couchbase Server Enterprise Edition) first.

> ifdef::show_edition[:page-edition: Enterprise Edition]

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  ifdef::show_edition[:page-edition: Couchbase Server Enterprise Edition]
  ```
  
  Using these Vale rules:
  * Couchbase.CouchbaseServerEnterprise (error) - Use the full product name (Couchbase Server Enterprise Edition) first.

> ifdef::show_edition[:page-edition: Enterprise Edition]

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  ifdef::show_edition[:page-edition: Couchbase Server Enterprise Edition]
  ```
  
  Using these Vale rules:
  * Couchbase.CouchbaseServerEnterprise (error) - Use the full product name (Couchbase Server Enterprise Edition) first.

> ifdef::show_edition[:page-edition: Enterprise Edition]

Reviewer: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
        ifdef::show_edition[:page-edition: Couchbase Server Enterprise Edition]
  ```
  
  Using these Vale rules:
  * Couchbase.CouchbaseServerEnterprise (error) - Use the full product name (Couchbase Server Enterprise Edition) first.

:page-role:
:description: Couchbase Lite database peer-to-peer sync- integrate a custom built listener

:source-language: Java

:source-language: Kotlin

// :param-name: kotlin
// :param-title: Android
// :param-module: android

[abstract]
--

> Description -- _{description}_ +

Author: Automated review comment from Vale and OpenAI:

  ```suggestion
  Description -- {description} +
  ```
  
  Using these Vale rules:
  * [Couchbase.Italics](https://docs.couchbase.com/styleguide/italic.html) (error) - Don&#x27;t use italics. Find another way to add emphasis.

> Description -- _{description}_ +

Author: Automated review comment from Vale and OpenAI using the gpt-4.1 model:

  ```suggestion
  Description -- {description} +
  ```
  
  Using these Vale rules:
  * [Couchbase.Italics](https://docs.couchbase.com/styleguide/italic.html) (error) - Don&#x27;t use italics. Find another way to add emphasis.

> Description -- _{description}_ +

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Description -- {description} +
  ```
  
  Using these Vale rules:
  * [Couchbase.Italics](https://docs.couchbase.com/styleguide/italic.html) (error) - Don&#x27;t use italics. Find another way to add emphasis.

> Description -- _{description}_ +

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Description -- {description} +
  ```
  
  Using these Vale rules:
  * [Couchbase.Italics](https://docs.couchbase.com/styleguide/italic.html) (error) - Don&#x27;t use italics. Find another way to add emphasis.

> Description -- _{description}_ +

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Description -- {description} +
  ```
  
  Using these Vale rules:
  * [Couchbase.Italics](https://docs.couchbase.com/styleguide/italic.html) (error) - Don&#x27;t use italics. Find another way to add emphasis.

> Description -- _{description}_ +

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Description -- {description} +
  ```
  
  Using these Vale rules:
  * [Couchbase.Italics](https://docs.couchbase.com/styleguide/italic.html) (error) - Don&#x27;t use italics. Find another way to add emphasis.

> Description -- _{description}_ +

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Description -- {description} +
  ```
  
  Using these Vale rules:
  * [Couchbase.Italics](https://docs.couchbase.com/styleguide/italic.html) (error) - Don&#x27;t use italics. Find another way to add emphasis.

Related Content -- https://docs.couchbase.com/mobile/{major}.{minor}.{maintenance-android}{empty}/couchbase-lite-android/[API Reference] | xref:android:p2psync-websocket.adoc[Peer-to-Peer] | xref:android:p2psync-custom.adoc[Integrate Custom Listener]
--

[#overview]
== Overview

> .Enterprise Edition only

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  .Couchbase Server Enterprise Edition only
  ```
  
  Using these Vale rules:
  * Couchbase.CouchbaseServerEnterprise (error) - Use the full product name (Couchbase Server Enterprise Edition) first.

> .Enterprise Edition only

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  .Couchbase Server Enterprise Edition only
  ```
  
  Using these Vale rules:
  * Couchbase.CouchbaseServerEnterprise (error) - Use the full product name (Couchbase Server Enterprise Edition) first.

> .Enterprise Edition only

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  .Couchbase Server Enterprise Edition only
  ```
  
  Using these Vale rules:
  * Couchbase.CouchbaseServerEnterprise (error) - Use the full product name (Couchbase Server Enterprise Edition) first.

> .Enterprise Edition only

Reviewer: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
        .Couchbase Server Enterprise Edition only
  ```
  
  Using these Vale rules:
  * Couchbase.CouchbaseServerEnterprise (error) - Use the full product name (Couchbase Server Enterprise Edition) first.

IMPORTANT: Peer-to-Peer Synchronization is an https://www.couchbase.com/products/editions[Enterprise Edition] feature.
You must purchase the Enterprise License, which includes official https://www.couchbase.com/support-policy[Couchbase Support].
To use it in production (also see the https://www.couchbase.com/licensing-and-support-faq[FAQ]).

This content covers how to integrate a custom __MessageEndpointListener__ solution with Couchbase Lite to handle the data transfer, which is the sending and receiving of data.

> Where applicable, we discuss how to integrate Couchbase Lite into the workflow.

Author: Automated review comment from Vale and OpenAI:

  ```suggestion
  Where applicable, this document discusses how to integrate Couchbase Lite into the workflow.
  ```
  
  Using these Vale rules:
  * [Google.We](https://developers.google.com/style/pronouns#personal-pronouns) (warning) - Try to avoid using first-person plural like &#x27;we&#x27;.

> Where applicable, we discuss how to integrate Couchbase Lite into the workflow.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Where applicable, this document discusses how to integrate Couchbase Lite into the workflow.
  ```
  
  Using these Vale rules:
  * [Google.We](https://developers.google.com/style/pronouns#personal-pronouns) (warning) - Try to avoid using first-person plural like &#x27;we&#x27;.

> Where applicable, we discuss how to integrate Couchbase Lite into the workflow.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Where applicable, this document discusses how to integrate Couchbase Lite into the workflow.
  ```
  
  Using these Vale rules:
  * [Google.We](https://developers.google.com/style/pronouns#personal-pronouns) (warning) - Try to avoid using first-person plural like &#x27;we&#x27;.

> Where applicable, we discuss how to integrate Couchbase Lite into the workflow.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Where applicable, this document discusses how to integrate Couchbase Lite into the workflow.
  ```
  
  Using these Vale rules:
  * [Google.We](https://developers.google.com/style/pronouns#personal-pronouns) (warning) - Try to avoid using first-person plural like &#x27;we&#x27;.

> Where applicable, we discuss how to integrate Couchbase Lite into the workflow.

Reviewer: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
        Where applicable, this document discusses how to integrate Couchbase Lite into the workflow.
  ```
  
  Using these Vale rules:
  * [Google.We](https://developers.google.com/style/pronouns#personal-pronouns) (warning) - Try to avoid using first-person plural like &#x27;we&#x27;.

The following sections describe a typical Peer-to-Peer workflow.

[#peer-discovery]
== Peer Discovery

Peer discovery is the first step.

> The communication framework will generally include a Peer discovery API for devices to advertise themselves on the network and to browse for other Peers.

Author: Automated review comment from Vale and OpenAI:

  ```suggestion
  The communication framework generally includes a Peer discovery API for devices to advertise themselves on the network and to browse for other Peers.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> The communication framework will generally include a Peer discovery API for devices to advertise themselves on the network and to browse for other Peers.

Author: Automated review comment from Vale and OpenAI using the gpt-4.1 model:

  ```suggestion
  The communication framework generally includes a Peer discovery API for devices to advertise themselves on the network and to browse for other Peers.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> The communication framework will generally include a Peer discovery API for devices to advertise themselves on the network and to browse for other Peers.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  The communication framework generally includes a Peer discovery API for devices to advertise themselves on the network and to browse for other Peers.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> The communication framework will generally include a Peer discovery API for devices to advertise themselves on the network and to browse for other Peers.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  The communication framework generally includes a Peer discovery API for devices to advertise themselves on the network and to browse for other Peers.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> The communication framework will generally include a Peer discovery API for devices to advertise themselves on the network and to browse for other Peers.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  The communication framework generally includes a Peer discovery API for devices to advertise themselves on the network and to browse for other Peers.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> The communication framework will generally include a Peer discovery API for devices to advertise themselves on the network and to browse for other Peers.

Reviewer: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
        The communication framework generally includes a Peer discovery API for devices to advertise themselves on the network and to browse for other Peers.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

[source]
----
include::android:example$codesnippet_collection.kt[tag=listener,indent=0]
----

[#peer-selection-and-connection-setup]
== Peer Selection and Connection Setup

Once a Peer device is found, the application code must decide whether it should establish a connection with that Peer.
This step includes inviting a Peer to a session and Peer authentication.

This is handled by the Communication Framework.

image::ROOT:selection.png[]

Once the remote Peer has been authenticated, the next step is to connect with that Peer and initialize the Message Endpoint API.

[#replication-setup]
== Replication Setup

image::ROOT:connection.png[]

[#active-peer-2]
=== Active Peer

When the connection is established, the active Peer must instantiate a `MessageEndpoint` object corresponding to the remote Peer.

[source]
----
include::android:example$codesnippet_collection.kt[tag=message-endpoint,indent=0]
----

The `MessageEndpoint` initializer takes the following arguments.

. `uid`: a unique ID that represents the remote active Peer.
. `target`: This represents the remote passive Peer and could be any suitable representation of the remote Peer.

> It could be an Id, URL etc.

Author: Automated review comment from Vale and OpenAI using the gpt-4.1 model:

  ```suggestion
  It could be an Id, URL and so on.
  ```
  
  Using these Vale rules:
  * [Couchbase.Terminology](https://docs.couchbase.com/styleguide/word-list.html) (warning) - Use &#x27;and so on&#x27;, not &#x27;etc&#x27;

> It could be an Id, URL etc.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  It could be an Id, URL and so on.
  ```
  
  Using these Vale rules:
  * [Couchbase.Terminology](https://docs.couchbase.com/styleguide/word-list.html) (warning) - Use &#x27;and so on&#x27;, not &#x27;etc&#x27;

> It could be an Id, URL etc.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  It could be an Id, URL and so on.
  ```
  
  Using these Vale rules:
  * [Couchbase.Terminology](https://docs.couchbase.com/styleguide/word-list.html) (warning) - Use &#x27;and so on&#x27;, not &#x27;etc&#x27;

> It could be an Id, URL etc.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  It could be an Id, URL and so on.
  ```
  
  Using these Vale rules:
  * [Couchbase.Terminology](https://docs.couchbase.com/styleguide/word-list.html) (warning) - Use &#x27;and so on&#x27;, not &#x27;etc&#x27;

> It could be an Id, URL etc.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  It could be an Id, URL and so on.
  ```
  
  Using these Vale rules:
  * [Couchbase.Terminology](https://docs.couchbase.com/styleguide/word-list.html) (warning) - Use &#x27;and so on&#x27;, not &#x27;etc&#x27;

> It could be an Id, URL etc.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  It could be an Id, URL and so on.
  ```
  
  Using these Vale rules:
  * [Couchbase.Terminology](https://docs.couchbase.com/styleguide/word-list.html) (warning) - Use &#x27;and so on&#x27;, not &#x27;etc&#x27;

> It could be an Id, URL etc.

Reviewer: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
        It could be an Id, URL and so on.
  ```
  
  Using these Vale rules:
  * [Couchbase.Terminology](https://docs.couchbase.com/styleguide/word-list.html) (warning) - Use &#x27;and so on&#x27;, not &#x27;etc&#x27;

If using the MultiPeerConnectivity Framework, this could be the MCPeerID.
. `protocolType`: specifies the kind of transport you intend to implement.

> There are two options.

Author: Automated review comment from Vale and OpenAI:

  ```suggestion
  Two options exist.
  ```
  
  Using these Vale rules:
  * write-good.ThereIs (error) - Don&#x27;t start a sentence with &#x27;There are&#x27;. 
  * Couchbase.ThereIs (error) - Don&#x27;t start a sentence with &quot;there&quot; + a form of &quot;to be.&quot;

> There are two options.

Author: Automated review comment from Vale and OpenAI using the gpt-4.1 model:

  ```suggestion
  Two options are available.
  ```
  
  Using these Vale rules:
  * write-good.ThereIs (error) - Don&#x27;t start a sentence with &#x27;There are&#x27;. 
  * Couchbase.ThereIs (error) - Don&#x27;t start a sentence with &quot;there&quot; + a form of &quot;to be.&quot;

> There are two options.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Two options exist.
  ```
  
  Using these Vale rules:
  * write-good.ThereIs (error) - Don&#x27;t start a sentence with &#x27;There are&#x27;. 
  * Couchbase.ThereIs (error) - Don&#x27;t start a sentence with &quot;there&quot; + a form of &quot;to be.&quot;

> There are two options.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Two options exist.
  ```
  
  Using these Vale rules:
  * Couchbase.ThereIs (error) - Don&#x27;t start a sentence with &quot;there&quot; + a form of &quot;to be.&quot; 
  * write-good.ThereIs (error) - Don&#x27;t start a sentence with &#x27;There are&#x27;.

> There are two options.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Two options exist.
  ```
  
  Using these Vale rules:
  * Couchbase.ThereIs (error) - Don&#x27;t start a sentence with &quot;there&quot; + a form of &quot;to be.&quot; 
  * write-good.ThereIs (error) - Don&#x27;t start a sentence with &#x27;There are&#x27;.

> There are two options.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Two options exist.
  ```
  
  Using these Vale rules:
  * write-good.ThereIs (error) - Don&#x27;t start a sentence with &#x27;There are&#x27;. 
  * Couchbase.ThereIs (error) - Don&#x27;t start a sentence with &quot;there&quot; + a form of &quot;to be.&quot;

> There are two options.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Two options exist.
  ```
  
  Using these Vale rules:
  * write-good.ThereIs (error) - Don&#x27;t start a sentence with &#x27;There are&#x27;. 
  * Couchbase.ThereIs (error) - Don&#x27;t start a sentence with &quot;there&quot; + a form of &quot;to be.&quot;

> There are two options.

Reviewer: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
        Two options exist.
  ```
  
  Using these Vale rules:
  * write-good.ThereIs (error) - Don&#x27;t start a sentence with &#x27;There are&#x27;. 
  * Couchbase.ThereIs (error) - Don&#x27;t start a sentence with &quot;there&quot; + a form of &quot;to be.&quot;

>  ** The default (`MessageStream`) means that you want to "send a series of messages", or in other words the Communication Framework will control the formatting of messages so that there are clear boundaries between messages.

Author: Automated review comment from Vale and OpenAI:

  ```suggestion
  ** The default (`MessageStream`) means that you want to "send a series of messages," or in other words the Communication Framework controls the formatting of messages so that there are clear boundaries between messages.
  ```
  
  Using these Vale rules:
  * [Google.Quotes](https://developers.google.com/style/quotation-marks) (error) - Commas and periods go inside quotation marks. 
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

>  ** The default (`MessageStream`) means that you want to "send a series of messages", or in other words the Communication Framework will control the formatting of messages so that there are clear boundaries between messages.

Author: Automated review comment from Vale and OpenAI using the gpt-4.1 model:

  ```suggestion
   ** The default (`MessageStream`) means that you want to "send a series of messages," or in other words the Communication Framework controls the formatting of messages so that there are clear boundaries between messages.
  ```
  
  Using these Vale rules:
  * [Google.Quotes](https://developers.google.com/style/quotation-marks) (error) - Commas and periods go inside quotation marks. 
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

>  ** The default (`MessageStream`) means that you want to "send a series of messages", or in other words the Communication Framework will control the formatting of messages so that there are clear boundaries between messages.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  ** The default (`MessageStream`) means that you want to "send a series of messages," or in other words the Communication Framework controls the formatting of messages so that there are clear boundaries between messages.
  ```
  
  Using these Vale rules:
  * [Google.Quotes](https://developers.google.com/style/quotation-marks) (error) - Commas and periods go inside quotation marks. 
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

>  ** The default (`MessageStream`) means that you want to "send a series of messages", or in other words the Communication Framework will control the formatting of messages so that there are clear boundaries between messages.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  ** The default (`MessageStream`) means that you want to "send a series of messages," or in other words the Communication Framework controls the formatting of messages so that there are clear boundaries between messages.
  ```
  
  Using these Vale rules:
  * [Google.Quotes](https://developers.google.com/style/quotation-marks) (error) - Commas and periods go inside quotation marks. 
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

>  ** The default (`MessageStream`) means that you want to "send a series of messages", or in other words the Communication Framework will control the formatting of messages so that there are clear boundaries between messages.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  ** The default (`MessageStream`) means that you want to "send a series of messages," or in other words the Communication Framework controls the formatting of messages so that there are clear boundaries between messages.
  ```
  
  Using these Vale rules:
  * [Google.Quotes](https://developers.google.com/style/quotation-marks) (error) - Commas and periods go inside quotation marks. 
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

>  ** The default (`MessageStream`) means that you want to "send a series of messages", or in other words the Communication Framework will control the formatting of messages so that there are clear boundaries between messages.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  ** The default (`MessageStream`) means that you want to "send a series of messages," or in other words the Communication Framework controls the formatting of messages so that there are clear boundaries between messages.
  ```
  
  Using these Vale rules:
  * [Google.Quotes](https://developers.google.com/style/quotation-marks) (error) - Commas and periods go inside quotation marks. 
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

>  ** The alternative (`ByteStream`) means that you just want to send raw bytes over the stream and Couchbase should format for you to ensure that messages get delivered in full.

Author: Automated review comment from Vale and OpenAI using the gpt-4.1 model:

  ```suggestion
   ** The alternative (`ByteStream`) means that you just want to send raw bytes over the stream and Couchbase should format for you to make sure that messages get delivered in full.
  ```
  
  Using these Vale rules:
  * [Couchbase.Terminology](https://docs.couchbase.com/styleguide/word-list.html) (warning) - Use &#x27;make sure|verify&#x27;, not &#x27;ensure&#x27;

>  ** The alternative (`ByteStream`) means that you just want to send raw bytes over the stream and Couchbase should format for you to ensure that messages get delivered in full.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  ** The alternative (`ByteStream`) means that you just want to send raw bytes over the stream and Couchbase should format for you to make sure that messages get delivered in full.
  ```
  
  Using these Vale rules:
  * [Couchbase.Terminology](https://docs.couchbase.com/styleguide/word-list.html) (warning) - Use &#x27;make sure|verify&#x27;, not &#x27;ensure&#x27;

>  ** The alternative (`ByteStream`) means that you just want to send raw bytes over the stream and Couchbase should format for you to ensure that messages get delivered in full.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  ** The alternative (`ByteStream`) means that you just want to send raw bytes over the stream and Couchbase should format for you to make sure that messages get delivered in full.
  ```
  
  Using these Vale rules:
  * [Couchbase.Terminology](https://docs.couchbase.com/styleguide/word-list.html) (warning) - Use &#x27;make sure|verify&#x27;, not &#x27;ensure&#x27;

>  ** The alternative (`ByteStream`) means that you just want to send raw bytes over the stream and Couchbase should format for you to ensure that messages get delivered in full.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  ** The alternative (`ByteStream`) means that you just want to send raw bytes over the stream and Couchbase should format for you to make sure that messages get delivered in full.
  ```
  
  Using these Vale rules:
  * [Couchbase.Terminology](https://docs.couchbase.com/styleguide/word-list.html) (warning) - Use &#x27;make sure|verify&#x27;, not &#x27;ensure&#x27;

>  ** The alternative (`ByteStream`) means that you just want to send raw bytes over the stream and Couchbase should format for you to ensure that messages get delivered in full.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  ** The alternative (`ByteStream`) means that you just want to send raw bytes over the stream and Couchbase should format for you to make sure that messages get delivered in full.
  ```
  
  Using these Vale rules:
  * [Couchbase.Terminology](https://docs.couchbase.com/styleguide/word-list.html) (warning) - Use &#x27;make sure|verify&#x27;, not &#x27;ensure&#x27;

>  ** The alternative (`ByteStream`) means that you just want to send raw bytes over the stream and Couchbase should format for you to ensure that messages get delivered in full.

Reviewer: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
         ** The alternative (`ByteStream`) means that you just want to send raw bytes over the stream and Couchbase should format for you to make sure that messages get delivered in full.
  ```
  
  Using these Vale rules:
  * [Couchbase.Terminology](https://docs.couchbase.com/styleguide/word-list.html) (warning) - Use &#x27;make sure|verify&#x27;, not &#x27;ensure&#x27;

+

> Typically, the Communication Framework will handle message assembly and disassembly so you would use the `MessageType` option in most cases.

Author: Automated review comment from Vale and OpenAI using the gpt-4.1 model:

  ```suggestion
  Typically, the Communication Framework handles message assembly and disassembly so you would use the `MessageType` option in most cases.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> Typically, the Communication Framework will handle message assembly and disassembly so you would use the `MessageType` option in most cases.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Typically, the Communication Framework handles message assembly and disassembly so you would use the `MessageType` option in most cases.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> Typically, the Communication Framework will handle message assembly and disassembly so you would use the `MessageType` option in most cases.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Typically, the Communication Framework handles message assembly and disassembly so you would use the `MessageType` option in most cases.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> Typically, the Communication Framework will handle message assembly and disassembly so you would use the `MessageType` option in most cases.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Typically, the Communication Framework handles message assembly and disassembly so you would use the `MessageType` option in most cases.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> Typically, the Communication Framework will handle message assembly and disassembly so you would use the `MessageType` option in most cases.

Reviewer: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
        Typically, the Communication Framework handles message assembly and disassembly so you would use the `MessageType` option in most cases.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> . `delegate`: the delegate that will implement the `MessageEndpointDelegate` protocol, which is a factory for `MessageEndpointConnection`.

Author: Automated review comment from Vale and OpenAI:

  ```suggestion
  .`delegate`: the delegate that implements the `MessageEndpointDelegate` protocol, which is a factory for `MessageEndpointConnection`.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> . `delegate`: the delegate that will implement the `MessageEndpointDelegate` protocol, which is a factory for `MessageEndpointConnection`.

Author: Automated review comment from Vale and OpenAI using the gpt-4.1 model:

  ```suggestion
  . `delegate`: the delegate that implements the `MessageEndpointDelegate` protocol, which is a factory for `MessageEndpointConnection`.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> . `delegate`: the delegate that will implement the `MessageEndpointDelegate` protocol, which is a factory for `MessageEndpointConnection`.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  . `delegate`: the delegate that implements the `MessageEndpointDelegate` protocol, which is a factory for `MessageEndpointConnection`.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> . `delegate`: the delegate that will implement the `MessageEndpointDelegate` protocol, which is a factory for `MessageEndpointConnection`.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  . `delegate`: the delegate that implements the `MessageEndpointDelegate` protocol, which is a factory for `MessageEndpointConnection`.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> . `delegate`: the delegate that will implement the `MessageEndpointDelegate` protocol, which is a factory for `MessageEndpointConnection`.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  . `delegate`: the delegate that implements the `MessageEndpointDelegate` protocol, which is a factory for `MessageEndpointConnection`.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> . `delegate`: the delegate that will implement the `MessageEndpointDelegate` protocol, which is a factory for `MessageEndpointConnection`.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  . `delegate`: the delegate that implements the `MessageEndpointDelegate` protocol, which is a factory for `MessageEndpointConnection`.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> . `delegate`: the delegate that will implement the `MessageEndpointDelegate` protocol, which is a factory for `MessageEndpointConnection`.

Reviewer: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
        . `delegate`: the delegate that implements the `MessageEndpointDelegate` protocol, which is a factory for `MessageEndpointConnection`.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

Then, a `Replicator` is instantiated with the initialized `MessageEndpoint` as the target.

[source]
----
include::android:example$codesnippet_collection.kt[tag=message-endpoint-replicator,indent=0]
----

> Next, Couchbase Lite will call back the application code through the `MessageEndpointDelegate.createConnection` interface method.

Author: Automated review comment from Vale and OpenAI:

  ```suggestion
  Next, Couchbase Lite calls back the application code through the `MessageEndpointDelegate.createConnection` interface method.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> Next, Couchbase Lite will call back the application code through the `MessageEndpointDelegate.createConnection` interface method.

Author: Automated review comment from Vale and OpenAI using the gpt-4.1 model:

  ```suggestion
  Next, Couchbase Lite calls back the application code through the `MessageEndpointDelegate.createConnection` interface method.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> Next, Couchbase Lite will call back the application code through the `MessageEndpointDelegate.createConnection` interface method.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Next, Couchbase Lite calls back the application code through the `MessageEndpointDelegate.createConnection` interface method.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> Next, Couchbase Lite will call back the application code through the `MessageEndpointDelegate.createConnection` interface method.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Next, Couchbase Lite calls back the application code through the `MessageEndpointDelegate.createConnection` interface method.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> Next, Couchbase Lite will call back the application code through the `MessageEndpointDelegate.createConnection` interface method.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Next, Couchbase Lite calls back the application code through the `MessageEndpointDelegate.createConnection` interface method.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> Next, Couchbase Lite will call back the application code through the `MessageEndpointDelegate.createConnection` interface method.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Next, Couchbase Lite calls back the application code through the `MessageEndpointDelegate.createConnection` interface method.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> Next, Couchbase Lite will call back the application code through the `MessageEndpointDelegate.createConnection` interface method.

Reviewer: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
        Next, Couchbase Lite calls back the application code through the `MessageEndpointDelegate.createConnection` interface method.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

When the application receives the callback, it must create an instance of `MessageEndpointConnection` and return it.

[source]
----
include::android:example$codesnippet_collection.kt[tag=create-connection,indent=0]
----

> Next, Couchbase Lite will call back the application code through the `MessageEndpointConnection.open` method.

Author: Automated review comment from Vale and OpenAI:

  ```suggestion
  Next, Couchbase Lite calls back the application code through the `MessageEndpointConnection.open` method.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> Next, Couchbase Lite will call back the application code through the `MessageEndpointConnection.open` method.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Next, Couchbase Lite calls back the application code through the `MessageEndpointConnection.open` method.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> Next, Couchbase Lite will call back the application code through the `MessageEndpointConnection.open` method.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Next, Couchbase Lite calls back the application code through the `MessageEndpointConnection.open` method.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> Next, Couchbase Lite will call back the application code through the `MessageEndpointConnection.open` method.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Next, Couchbase Lite calls back the application code through the `MessageEndpointConnection.open` method.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

[source]
----
include::android:example$codesnippet_collection.kt[tag=active-peer-open,indent=0]
----

The connection argument is then set on an instance variable.
The application code must keep track of every `ReplicatorConnection` associated with every `MessageEndpointConnection`.

The `MessageError` argument in the completion block specifies whether the error is recoverable or not.

> If it is a recoverable error, the replicator will begin a retry process, creating a new `MessageEndpointConnection` instance.

Author: Automated review comment from Vale and OpenAI:

  ```suggestion
  If it's a recoverable error, the replicator begins a retry process, creating a new `MessageEndpointConnection` instance.
  ```
  
  Using these Vale rules:
  * [Couchbase.Contractions](https://docs.couchbase.com/styleguide/contractions.html) (error) - Use &#x27;it&#x27;s&#x27; instead of &#x27;it is&#x27;. 
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> If it is a recoverable error, the replicator will begin a retry process, creating a new `MessageEndpointConnection` instance.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  If it's a recoverable error, the replicator begins a retry process, creating a new `MessageEndpointConnection` instance.
  ```
  
  Using these Vale rules:
  * [Couchbase.Contractions](https://docs.couchbase.com/styleguide/contractions.html) (error) - Use &#x27;it&#x27;s&#x27; instead of &#x27;it is&#x27;. 
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> If it is a recoverable error, the replicator will begin a retry process, creating a new `MessageEndpointConnection` instance.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  If it's a recoverable error, the replicator begins a retry process, creating a new `MessageEndpointConnection` instance.
  ```
  
  Using these Vale rules:
  * [Couchbase.Contractions](https://docs.couchbase.com/styleguide/contractions.html) (error) - Use &#x27;it&#x27;s&#x27; instead of &#x27;it is&#x27;. 
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> If it is a recoverable error, the replicator will begin a retry process, creating a new `MessageEndpointConnection` instance.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  If it's a recoverable error, the replicator begins a retry process, creating a new `MessageEndpointConnection` instance.
  ```
  
  Using these Vale rules:
  * [Couchbase.Contractions](https://docs.couchbase.com/styleguide/contractions.html) (error) - Use &#x27;it&#x27;s&#x27; instead of &#x27;it is&#x27;. 
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

[#passive-peer-2]
=== Passive Peer

After connection establishment on the Passive Peer, the first step is to initialize a new `MessageEndpointConnection` and pass it to the listener.
This message tells the listener to accept incoming data from that Peer.

[source]
----
include::android:example$codesnippet_collection.kt[tag=advertizer-accept,indent=0]
----

`messageEndpointListener` is the instance of the `MessageEndpointListener` that was created in the first step (<<peer-discovery,Peer Discovery>>)

> Couchbase Lite will call the application code back through the `MessageEndpointConnection.open` method.

Author: Automated review comment from Vale and OpenAI using the gpt-4.1 model:

  ```suggestion
  Couchbase Lite calls the application code back through the `MessageEndpointConnection.open` method.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> Couchbase Lite will call the application code back through the `MessageEndpointConnection.open` method.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Couchbase Lite calls the application code back through the `MessageEndpointConnection.open` method.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> Couchbase Lite will call the application code back through the `MessageEndpointConnection.open` method.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Couchbase Lite calls the application code back through the `MessageEndpointConnection.open` method.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> Couchbase Lite will call the application code back through the `MessageEndpointConnection.open` method.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Couchbase Lite calls the application code back through the `MessageEndpointConnection.open` method.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> Couchbase Lite will call the application code back through the `MessageEndpointConnection.open` method.

Reviewer: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
        Couchbase Lite calls the application code back through the `MessageEndpointConnection.open` method.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

[source]
----
include::android:example$codesnippet_collection.kt[tag=passive-peer-open,indent=0]
----

The `connection` argument is then set on an instance variable.
The application code must keep track of every `ReplicatorConnection` associated with every `MessageEndpointConnection`.

At this point, the connection is established, and both Peers are ready to exchange data.

[#pushpull-replication]
== Push/Pull Replication

Typically, an application needs to send data and receive data.
The directionality of the replication could be any of the following.

* *Push only:* The data is pushed from the local database to the remote database.

* *Pull only:* The data is pulled from the remote database to the local database.

* *Push and Pull:* The data is exchanged both ways.

> Usually, the remote is a Sync Gateway database identified through a URL.

Author: Automated review comment from Vale and OpenAI:

  ```suggestion
  In most cases, the remote is a Sync Gateway database identified through a URL.
  ```
  
  Using these Vale rules:
  * [write-good.Weasel](https://en.wikipedia.org/wiki/Weasel_word) (warning) - &#x27;Usually&#x27; is a weasel word!

> Usually, the remote is a Sync Gateway database identified through a URL.

Author: Automated review comment from Vale and OpenAI using the gpt-4.1 model:

  ```suggestion
  The remote is typically a Sync Gateway database identified through a URL.
  ```
  
  Using these Vale rules:
  * [write-good.Weasel](https://en.wikipedia.org/wiki/Weasel_word) (warning) - &#x27;Usually&#x27; is a weasel word!

Author: isn't "typically" also a weasel?

> Usually, the remote is a Sync Gateway database identified through a URL.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Typically, the remote is a Sync Gateway database identified through a URL.
  ```
  
  Using these Vale rules:
  * [write-good.Weasel](https://en.wikipedia.org/wiki/Weasel_word) (warning) - &#x27;Usually&#x27; is a weasel word!

In Peer-to-Peer syncing, the remote is another Couchbase Lite database.

image::ROOT:replication.png[]

The replication lifecycle is handled through the `MessageEndpointConnection`.

[#active-peer-3]
=== Active Peer

When Couchbase Lite calls back the application code through the `MessageEndpointConnection.send` method, you should send that data to the other Peer using the communication framework.

[source]
----
include::android:example$codesnippet_collection.kt[tag=active-peer-send,indent=0]
----

Once the data is sent, call the completion block to acknowledge the completion.
You can use the `MessageError` in the completion block to specify whether the error is recoverable.

> If it is a recoverable error, the replicator will begin a retry process, creating a new `MessageEndpointConnection`.

Author: Automated review comment from Vale and OpenAI:

  ```suggestion
  If it's a recoverable error, the replicator begins a retry process, creating a new `MessageEndpointConnection`.
  ```
  
  Using these Vale rules:
  * [Couchbase.Contractions](https://docs.couchbase.com/styleguide/contractions.html) (error) - Use &#x27;it&#x27;s&#x27; instead of &#x27;it is&#x27;. 
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> If it is a recoverable error, the replicator will begin a retry process, creating a new `MessageEndpointConnection`.

Author: Automated review comment from Vale and OpenAI using the gpt-4.1 model:

  ```suggestion
  If it's a recoverable error, the replicator begins a retry process, creating a new `MessageEndpointConnection`.
  ```
  
  Using these Vale rules:
  * [Couchbase.Contractions](https://docs.couchbase.com/styleguide/contractions.html) (error) - Use &#x27;it&#x27;s&#x27; instead of &#x27;it is&#x27;. 
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> If it is a recoverable error, the replicator will begin a retry process, creating a new `MessageEndpointConnection`.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  If it's a recoverable error, the replicator begins a retry process, creating a new `MessageEndpointConnection`.
  ```
  
  Using these Vale rules:
  * [Couchbase.Contractions](https://docs.couchbase.com/styleguide/contractions.html) (error) - Use &#x27;it&#x27;s&#x27; instead of &#x27;it is&#x27;. 
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> If it is a recoverable error, the replicator will begin a retry process, creating a new `MessageEndpointConnection`.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  If it's a recoverable error, the replicator begins a retry process, creating a new `MessageEndpointConnection`.
  ```
  
  Using these Vale rules:
  * [Couchbase.Contractions](https://docs.couchbase.com/styleguide/contractions.html) (error) - Use &#x27;it&#x27;s&#x27; instead of &#x27;it is&#x27;. 
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> If it is a recoverable error, the replicator will begin a retry process, creating a new `MessageEndpointConnection`.

Reviewer: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
        If it's a recoverable error, the replicator begins a retry process, creating a new `MessageEndpointConnection`.
  ```
  
  Using these Vale rules:
  * [Couchbase.Contractions](https://docs.couchbase.com/styleguide/contractions.html) (error) - Use &#x27;it&#x27;s&#x27; instead of &#x27;it is&#x27;. 
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

When data is received from the passive Peer via the Communication Framework, you call the `ReplicatorConnection.receive` method.

[source]
----
include::android:example$codesnippet_collection.kt[tag=active-peer-receive,indent=0]
----

> The replication connection's `receive` method is called. Which then processes the data to persist to the local database.

Author: Automated review comment from Vale and OpenAI using the gpt-4.1 model:

  ```suggestion
  The replication connection's `receive` method is called.
Which then processes the data to persist to the local database.
  ```
  
  Using these Vale rules:
  * Couchbase.VentilatedProse (warning) - Use ventilated prose. Start every sentence on a new line.

> The replication connection's `receive` method is called. Which then processes the data to persist to the local database.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  The replication connection's `receive` method is called. 

Which then processes the data to persist to the local database.
  ```
  
  Using these Vale rules:
  * Couchbase.VentilatedProse (warning) - Use ventilated prose. Start every sentence on a new line.

[#passive-peer-3]
=== Passive Peer

As in the case of the active Peer, the passive Peer must implement the `MessageEndpointConnection.send` method to send data to the other Peer.

[source]
----
include::android:example$codesnippet_collection.kt[tag=passive-peer-send,indent=0]
----

Once the data is sent, call the completion block to acknowledge the completion.
You can use the `MessageError` in the completion block to specify whether the error is recoverable.

> If it is a recoverable error, the replicator will begin a retry process, creating a new `MessageEndpointConnection`.

Author: Automated review comment from Vale and OpenAI using the gpt-4.1 model:

  ```suggestion
  If it's a recoverable error, the replicator begins a retry process, creating a new `MessageEndpointConnection`.
  ```
  
  Using these Vale rules:
  * [Couchbase.Contractions](https://docs.couchbase.com/styleguide/contractions.html) (error) - Use &#x27;it&#x27;s&#x27; instead of &#x27;it is&#x27;. 
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> If it is a recoverable error, the replicator will begin a retry process, creating a new `MessageEndpointConnection`.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  If it's a recoverable error, the replicator begins a retry process, creating a new `MessageEndpointConnection`.
  ```
  
  Using these Vale rules:
  * [Couchbase.Contractions](https://docs.couchbase.com/styleguide/contractions.html) (error) - Use &#x27;it&#x27;s&#x27; instead of &#x27;it is&#x27;. 
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> If it is a recoverable error, the replicator will begin a retry process, creating a new `MessageEndpointConnection`.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  If it's a recoverable error, the replicator begins a retry process, creating a new `MessageEndpointConnection`.
  ```
  
  Using these Vale rules:
  * [Couchbase.Contractions](https://docs.couchbase.com/styleguide/contractions.html) (error) - Use &#x27;it&#x27;s&#x27; instead of &#x27;it is&#x27;. 
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

When data is received from the active Peer via the Communication Framework, you call the `ReplicatorConnection.receive` method.

[source]
----
include::android:example$codesnippet_collection.kt[tag=passive-peer-receive,indent=0]
----

[#connection-teardown]
== Connection Teardown

When a Peer disconnects from a Peer-to-Peer network, all connected Peers are notified.
The disconnect notification is a good opportunity to close and remove a replication connection.
The steps to Teardown the connection are slightly different depending on whether the active or passive Peer disconnects first.

> We will cover each case below.

Author: Automated review comment from Vale and OpenAI:

  ```suggestion
  This document covers each case.
  ```
  
  Using these Vale rules:
  * [Google.We](https://developers.google.com/style/pronouns#personal-pronouns) (warning) - Try to avoid using first-person plural like &#x27;We&#x27;. 
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;. 
  * Couchbase.DirectionalLanguage (warning) - Avoid directional language like &#x27;below&#x27;.

> We will cover each case below.

Author: Automated review comment from Vale and OpenAI using the gpt-4.1 model:

  ```suggestion
  Each case is covered in this documentation.
  ```
  
  Using these Vale rules:
  * [Google.We](https://developers.google.com/style/pronouns#personal-pronouns) (warning) - Try to avoid using first-person plural like &#x27;We&#x27;. 
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;. 
  * Couchbase.DirectionalLanguage (warning) - Avoid directional language like &#x27;below&#x27;.

> We will cover each case below.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  This document covers each case.
  ```
  
  Using these Vale rules:
  * [Google.We](https://developers.google.com/style/pronouns#personal-pronouns) (warning) - Try to avoid using first-person plural like &#x27;We&#x27;. 
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;. 
  * Couchbase.DirectionalLanguage (warning) - Avoid directional language like &#x27;below&#x27;.

> We will cover each case below.

Reviewer: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
        This document covers each case.
  ```
  
  Using these Vale rules:
  * [Google.We](https://developers.google.com/style/pronouns#personal-pronouns) (warning) - Try to avoid using first-person plural like &#x27;We&#x27;. 
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;. 
  * Couchbase.DirectionalLanguage (warning) - Avoid directional language like &#x27;below&#x27;.

[#initiated-by-active-peer]
=== Initiated by Active Peer

image::ROOT:dis-active.png[]

[#active-peer-4]
=== Active Peer

When an active Peer disconnects, it must call the `ReplicatorConnection.close` method.

[source]
----
include::android:example$codesnippet_collection.kt[tag=active-replicator-close,indent=0]
----

> Then, Couchbase Lite will call back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.

Author: Automated review comment from Vale and OpenAI:

  ```suggestion
  Then, Couchbase Lite calls back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> Then, Couchbase Lite will call back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.

Author: Automated review comment from Vale and OpenAI using the gpt-4.1 model:

  ```suggestion
  Then, Couchbase Lite calls back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> Then, Couchbase Lite will call back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Then, Couchbase Lite calls back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> Then, Couchbase Lite will call back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.

Reviewer: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
        Then, Couchbase Lite calls back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

[source]
----
include::android:example$codesnippet_collection.kt[tag=active-peer-close,indent=0]
----

[#passive-peer-4]
=== Passive Peer

When the passive Peer receives the corresponding disconnect notification from the Communication Framework, it must call the `ReplicatorConnection.close` method.

[source]
----
include::android:example$codesnippet_collection.kt[tag=passive-replicator-close,indent=0]
----

> Then, Couchbase Lite will call back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.

Author: Automated review comment from Vale and OpenAI:

  ```suggestion
  Then, Couchbase Lite calls back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> Then, Couchbase Lite will call back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.

Author: Automated review comment from Vale and OpenAI using the gpt-4.1 model:

  ```suggestion
  Then, Couchbase Lite calls back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> Then, Couchbase Lite will call back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Then, Couchbase Lite calls back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> Then, Couchbase Lite will call back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.

Reviewer: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
        Then, Couchbase Lite calls back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

[source]
----
include::android:example$codesnippet_collection.kt[tag=passive-peer-close,indent=0]
----

[#initiated-by-passive-peer]
=== Initiated by Passive Peer

image::ROOT:dis-passive.png[]

[#passive-peer-5]
=== Passive Peer

When the passive disconnects, it must class the `MessageEndpointListener.closeAll` method.

[source]
----
include::android:example$codesnippet_collection.kt[tag=passive-stop-listener,indent=0]
----

> Then, Couchbase Lite will call back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.

Author: Automated review comment from Vale and OpenAI:

  ```suggestion
  Then, Couchbase Lite calls back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> Then, Couchbase Lite will call back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.

Author: Automated review comment from Vale and OpenAI using the gpt-4.1 model:

  ```suggestion
  Then, Couchbase Lite calls back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> Then, Couchbase Lite will call back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Then, Couchbase Lite calls back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

[source]
----
include::android:example$codesnippet_collection.kt[tag=passive-peer-close,indent=0]
----

[#active-peer-5]
=== Active Peer

When the active Peer receives the corresponding disconnect notification from the Communication Framework, it must call the `ReplicatorConnection.close` method.

[source]
----
include::android:example$codesnippet_collection.kt[tag=active-replicator-close,indent=0]
----

> Then, Couchbase Lite will call back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.

Author: Automated review comment from Vale and OpenAI using the gpt-4.1 model:

  ```suggestion
  Then, Couchbase Lite calls back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> Then, Couchbase Lite will call back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.

Author: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
  Then, Couchbase Lite calls back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.

> Then, Couchbase Lite will call back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.

Reviewer: Automated review comment from Vale and OpenAI using the gpt-4o-mini model:

  ```suggestion
        Then, Couchbase Lite calls back your code through the `MessageEndpointConnection.close` to allow the application to disconnect with the Communication Framework.
  ```
  
  Using these Vale rules:
  * [Google.Will](https://developers.google.com/style/tense) (warning) - Avoid using &#x27;will&#x27;.
